src/librustc_resolve/lib.rs

```diff
@@ -1685,6 +1684,12 @@ pub struct Resolver<'a> {
     local_macro_def_scopes: FxHashMap<NodeId, Module<'a>>,
     unused_macros: NodeMap<Span>,
     proc_macro_stubs: NodeSet,
+    /// Some built-in derives mark items they are applied to so they are treated specially later.
+    /// Derive macros cannot modify the item themselves and have to store the markers in the global
+    /// context, so they attach the markers to derive container IDs using this resolver table.
+    /// FIXME: Find a way for `PartialEq` and `Eq` to emulate `#[structural_match]`
+    /// by marking the produced impls rather than the original items.
```

Comment:
I would think the better way to do this would be a perma unstable trait:

```rust
#[lang = "structural_match"]
unsafe trait Structural {}
```

for which implementations are generated.

Reply:
Yeah, something like that would also work.
`PartialEq` produces `impl Structural`, `rustc_mir` somehow checks for `T: Structural + Eq`.

Reply:
cc also @varkor

This could be useful for const generics (https://github.com/rust-lang/rust/issues/60286), e.g. `fn foo<T: Structural, const A: T>();`.